qa/python_models/dlpack_test/model.py

```diff
@@ -99,28 +119,172 @@ def test_dlpack_gpu_tensors(self):
             str(e.exception) ==
             'Tensor is stored in GPU and cannot be converted to NumPy.')
+    
+    def test_dlpack_cpu_numpy(self):
+        # Check compatibiity of PbTensor DLPack implementation
+        # with numpy
+        pytorch_tensor = torch.rand([100], dtype=torch.float16,
+                                    device='cpu') * 100
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
+        numpy_tensor_dlpack = np.from_dlpack(pb_tensor)
+        self.assertTrue(np.array_equal(
+            numpy_tensor_dlpack, pytorch_tensor.numpy()))
+        # Make sure that `__dlpack_device__` works as expected
+        self.assertTrue(pb_tensor.is_cpu())
+        self.assertFalse(pytorch_tensor.is_cuda)
+        self.assertEqual(
+            pb_tensor.__dlpack_device__(), pytorch_tensor.__dlpack_device__())
+    
+    def test_bool_datatype(self):
+        # [FIXME] pass bool_array directly to `pb_utils.Tensor.from_dlpack`,
+        # when numpy release supports DLPack bool type
+        bool_array = np.asarray([False, True])
+        bool_tensor = pb_utils.Tensor('tensor', bool_array)
+        bool_tensor_dlpack = pb_utils.Tensor.from_dlpack('tensor', bool_tensor)
+        self.assertTrue(
+            np.array_equal(bool_array, bool_tensor_dlpack.as_numpy()))
+    
+    def test_cuda_multi_stream(self):
+        # Test that external stream syncs with the default
+        # and pb_tensor has proper data
+        s1 = torch.cuda.Stream()
+        size = 5000
+        pytorch_tensor = torch.tensor([0,0,0,0], device='cuda')
+        expected_output = torch.tensor([2,2,2,2], device='cuda')
+        with torch.cuda.stream(s1):
+            matrix_a = torch.randn(size,size, device='cuda')
+            res = torch.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = torch.matmul(res,matrix_a)
+            pytorch_tensor += torch.tensor([2,2,2,2], device='cuda')
+        
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
+        pytorch_tensor_dlpack = from_dlpack(pb_tensor)
+        self.assertTrue(torch.equal(pytorch_tensor_dlpack, expected_output))
+    
+    def test_cuda_non_blocking_multi_stream(self):
+        # Test that external non-blocking stream syncs with the default stream
+        # and pb_tensor has proper data
+        non_blocking_stream = cp.cuda.Stream(non_blocking=True)
+        size = 5000
+        cupy_tensor = cp.array([0,0,0,0])
+        expected_output = cp.array([2,2,2,2])
+        with non_blocking_stream:
+            matrix_a = cp.random.rand(size,size)
+            res = cp.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = cp.matmul(res,matrix_a)
+            cupy_tensor += cp.array([2,2,2,2])
+        
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', cupy_tensor)
+        # Verify that non-blocking stream has no pending jobs left
+        self.assertTrue(non_blocking_stream.done)
```

Comment:
do we need to wait explicity - or that happends automatically with the `with` statement?

Reply:
with `with` statement we schedule work on external stream. `matmuls` take its time to do the work, I use it as a way to occupy the stream. Alternatively, I can write a kernel in pycuda, which makes the stream sleep, but I think `matmuls` do the same trick. In L210, we take a cupy tensor, and cupy's tensor's `__dlpack__` method should sync with our default stream through a cuda event - wait event scheme: https://github.com/cupy/cupy/blob/cbc69127e10c0188c17439a47c1ad58a9d32b2e0/cupy/_core/core.pyx#L320-L325

so after we consumed capsule, external stream should be done with computation. L212 asserts that.